python/sdk/merlin/observability/inference.py

```diff
@@ -1,111 +1,251 @@
+
+    """
+    Given an input dataframe, return a new dataframe with the necessary columns for observability,
+    along with a schema for the observability backend to parse the dataframe. In place changes might
+    be made to the input dataframe.
+    
+    :param df: Input dataframe.
+    :param observation_types: Types of observations to be included in the output dataframe.
+    :return: output dataframe
+    """
+
+    @abc.abstractmethod
+    def preprocess(
+        self, df: pd.DataFrame, observation_types: List[ObservationType]
+    ) -> pd.DataFrame:
+        raise NotImplementedError
+
+    """
+    Return a dictionary mapping the name of the prediction output column to its value type.
+    """
+
+    @abc.abstractmethod
+    def prediction_types(self) -> Dict[str, ValueType]:
+        raise NotImplementedError
 
 
 @dataclass_json
 @dataclass
-class RegressionOutput:
+class RegressionOutput(PredictionOutput):
+    """
+    Regression model prediction output schema.
+    Attributes:
+        prediction_score_column: Name of the column containing the prediction score.
+        actual_score_column: Name of the column containing the actual score.
+    """
+
     prediction_score_column: str
+    actual_score_column: str
 
-    @property
-    def column_types(self) -> Dict[str, ValueType]:
-        return {self.prediction_score_column: ValueType.FLOAT64}
+    def preprocess(
+        self, df: pd.DataFrame, observation_types: List[ObservationType]
+    ) -> pd.DataFrame:
+        return df
+
+    def prediction_types(self) -> Dict[str, ValueType]:
+        return {
+            self.prediction_score_column: ValueType.FLOAT64,
+            self.actual_score_column: ValueType.FLOAT64,
+        }
 
 
 @dataclass_json
 @dataclass
-class BinaryClassificationOutput:
+class BinaryClassificationOutput(PredictionOutput):
+    """
+    Binary classification model prediction output schema.
+
+    Attributes:
+        prediction_score_column: Name of the column containing the prediction score.
+            Prediction score must be between 0.0 and 1.0.
+        actual_label_column: Name of the column containing the actual class.
+        positive_class_label: Label for positive class.
+        negative_class_label: Label for negative class.
+        score_threshold: Score threshold for prediction to be considered as positive class.
+    """
+
     prediction_score_column: str
-    prediction_label_column: Optional[str] = None
+    actual_label_column: str
+    positive_class_label: str
+    negative_class_label: str
+    score_threshold: float = 0.5
 
     @property
-    def column_types(self) -> Dict[str, ValueType]:
-        column_types_mapping = {self.prediction_score_column: ValueType.FLOAT64}
-        if self.prediction_label_column is not None:
-            column_types_mapping[self.prediction_label_column] = ValueType.STRING
-        return column_types_mapping
-
-
-@dataclass_json
-@dataclass
-class MulticlassClassificationOutput:
```

Comment:
Why has `MulticlassClassificationOutput` been removed in this PR? Do we plan to address it later on only when there is a need?

Reply:
Multiclass model takes significantly more effort to implement compared to the other models, in terms of the preprocessing required for the input dataframe. And this effort is complicated by the fact that we do not really know how our users specify the input for multiclass models, both on online prediction and batch training data scenario. Therefore, we should address this later, when we have the clarity.